extensions/chathistory.md

```diff
@@ -0,0 +1,115 @@
+---
+title: IRCv3 chathistory extension
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: "Evan Magaliff"
+    period: "2017"
+    email: "evan@muffinmedic.net"
+  -
+    name: "Darren Whitlen"
+    period: "2018-2019"
+    email: "darren@kiwiirc.com"
+---
+## Description
+This document describes the format of the `chathistory` extension. This enables clients to request messages that were previously sent if they are still available on the server.
+
+The server as mentioned in this document may refer to either an IRC server or an IRC bouncer.
+
+## Implementation
+The `chathistory` extension uses the [chathistory][batch/chathistory] batch type and introduces a client command, `chathistory`.
+
+To fully support this extension, clients MUST support the [`batch`][batch], [`server-time`][server-time] and [`message-tags`][message-tags] capabilities.
+
+The `chathistory` capability MUST be negotiated. This allows the server and client to act differently when delivering message history on connection.
+ 
+An ISUPPORT token MUST be sent to the client to state the maximum number of messages a client can request in a single command, represented by an integer. `CHATHISTORY=50`. If `0`, the client SHOULD assume that there is no maximum number of messages.
+
+### `CHATHISTORY` Command
+`CHATHISTORY` content can be requested by the client by sending the `CHATHISTORY` command to the server. A `batch` MUST be returned by the server. If no content exists to return, an empty batch SHOULD be returned to avoid the client waiting for a reply and to indicate that no content is available.
+
+The `chathistory` command uses the following general syntax structure:
+
+    CHATHISTORY <subcommand> <target> <timestamp | msgid> <limit>
+
+The `target` parameter specifies a single buffer (channel or nickname) from which history SHOULD be retrieved. Any `timestamp` values or parameters MUST be in the format of `YYYY-MM-DDThh:mm:ss.sssZ`.
```

Comment:
In the situation where a client is disconnected to a bouncer offering chathistory. How can I know which nickname targets may have messaged me while I am disconnected? Someone I have never spoken to with the client may message me  while I am disconnected and the API offers no way to know which targets are available so the client can reconnect but it doesn’t know there are messages for a specific target.

This is less of a problem for channels because you likely get a JOIN event so you can retrieve the history afterwards. ZNC-Playback solves this problem by allowing wildcards here (although that introduces other problems of its own).

Reply:
I was thinking a special-purpose target like `*self` that refers to your own direct messages with all other clients, including your self-messages.

Reply:
This is resolved now by the `*` target (merged into the PR).